File: gifmaker.rb
#!/usr/bin/env ruby

require './scanner'

class GifMaker < Callback
    def initialize()
        super()
        throw "ffmpeg is required" if !is_ffmpeg_present()
        throw "ImageMagic convert is required" if !is_convert_present()
    end
    
    def is_ffmpeg_present()
        result = `which ffmpeg`
        return false if nil == result || result.empty?
        return true
    end
    private :is_ffmpeg_present

    def is_convert_present()
        result = `which convert`
        return false if nil == result || result.empty?
        return true
    end
    private :is_convert_present
    
    def timecode_string(seconds)
        hours = (seconds / (60 * 60)).to_i
        seconds = seconds % (60 * 60)
        minutes = (seconds / 60).to_i
        seconds = seconds % 60
        return sprintf("%02u:%02u:%02u.000", hours, minutes, seconds)
    end
    private :timecode_string

    def extract_frames(video_filename, extract_period, tmp_location)
        seconds = 0
        counter = 0
        while true
            timecode = timecode_string(seconds)
            filename = sprintf("%s/img%05d.jpg", tmp_location, counter)
            cmd = "ffmpeg -loglevel 16 -ss #{timecode} -i \"#{video_filename}\" -frames:v 1 \"#{filename}\""
            print("#{timecode}\r")
            rc = Kernel.system(cmd)
            break if (true != rc)
            break if (!File.exists?(filename))
            seconds += extract_period
            counter += 1
        end
    end
    private :extract_frames
    
    def load_frames(folder)
        Dir.new(folder).each { |filename|
            if (0 == filename.index('img') && filename.end_with?('.jpg'))
                @frames_original << File.expand_path(folder + '/' + filename);
            end
        }
    end
    private :load_frames
    
    def convert_frames()
        @frames_original.each { |filename|
            outfile = filename[0, filename.rindex('.')] + ".gif"
            @frames_resized << outfile
            cmd = "convert -resize 300x300 '" + filename + "' '" + outfile + "'"
            system(cmd)
        }
    end
    private :convert_frames
    
    def build_gif(output_file)
        cmd = "gifsicle --merge --delay 3 --loopcount=0 --optimize --colors 256"
        @frames_resized.sort.each { |file|
            cmd += " '" + file + "'"
        }
        cmd += " > '" + output_file + "'"
        system(cmd)
    end
    private :build_gif

    def callback(input_file)
        output_file = input_file[0, input_file.rindex('.')] + ".gif"
        return if !File.exists?(input_file)
        return if File.exists?(output_file)
        
        temp_folder = "/tmp/tagger/"
        `rm -rf #{temp_folder}`
        Dir.mkdir(temp_folder)

        print("#{input_file} => #{output_file}\n")
        @frames_original = Array.new
        @frames_resized = Array.new
        extract_frames(input_file, 1, temp_folder)
        load_frames(temp_folder)
        convert_frames()
        build_gif(output_file)
    end
end

File: gifmaker.py
"""GIF timelapse generation plugin for SecurityCamProcessor."""

import os

from PIL import Image
from scanner import Callback


class GifMaker(Callback):
    """Converts extracted video frames into an animated GIF timelapse."""

    def __init__(self) -> None:
        super().__init__()


    def needs_processing(self, input_file: str) -> bool:
        """Return True if a .gif file does not exist for this input."""
        output_file: str = os.path.splitext(input_file)[0] + ".gif"
        return not os.path.exists(output_file)

    def callback(self, input_file: str, frames: list[str]) -> None:
        """Resize frames and combine into an animated GIF."""
        output_file: str = os.path.splitext(input_file)[0] + ".gif"
        if not os.path.exists(input_file):
            return
        if os.path.exists(output_file):
            return

        print(f"{input_file} => {output_file}")
        resized_images: list[Image.Image] = self._resize_frames(sorted(frames))
        self._build_gif(output_file, resized_images)

    def _resize_frames(self, frames: list[str]) -> list[Image.Image]:
        """Open each JPEG frame, resize to 300x300 with LANCZOS, convert to palette mode."""
        images: list[Image.Image] = []
        for filename in frames:
            img: Image.Image = Image.open(filename)
            img = img.resize((300, 300), Image.Resampling.LANCZOS)
            img = img.convert("P", palette=Image.Palette.ADAPTIVE, colors=256)
            images.append(img)
        return images

    def _build_gif(self, output_file: str, images: list[Image.Image]) -> None:
        """Save list of PIL Images as animated GIF with 30ms delay and infinite loop."""
        if not images:
            return
        first: Image.Image = images[0]
        rest: list[Image.Image] = images[1:]
        first.save(
            output_file,
            save_all=True,
            append_images=rest,
            duration=30,
            loop=0,
        )
